rearranged page to fit upcoming blog post structure, added non-core EIPs (#16298)
ethereum/ethereum-org-website: public/content/roadmap/fusaka/index.md

## Improvements in Fusaka {#improvements-in-fusaka}

### Data availability & L2 scaling {#data-availability-and-l2-scaling}
### Scale blobs {#scale-blobs}

#### PeerDAS {#peerdas}

Specification: https://eips.ethereum.org/EIPS/eip-7594

> Resources: https://youtu.be/bONWd1x2TjQ?t=328 (dapplion on PeerDAS)

Author: thank you so much for the thorough review @wackerow !! moved the link to the bottom per your suggestion & also resolved the other comments

This keeps hardware and bandwidth requirements for nodes tenable while enabling blob scaling resulting in more scale with smaller fees for layer 2s.

**Resources**:
- [EIP-7594 technical specification](https://eips.ethereum.org/EIPS/eip-7594)
- [Main PeerDAS explainer page](https://ethereum.org/en/roadmap/fusaka/peerdas)
- [dapplion on PeerDAS](https://youtu.be/bONWd1x2TjQ?t=328)

> - [Academic: A Documentation of Ethereum’s PeerDAS](https://eprint.iacr.org/2024/1362.pdf)

Author: lgtm but "Lion" should be taken out (just leaving "dapplion" and "PeedDAS" corrected

Blob parameter only forks can be set by clients, similarly to other configuration like gas limit. Between major Ethereum upgrades, clients can agree to increase the `target` and `max` blobs to e.g. 9 and 12 and then node operators will update to take part in that tiny fork. These blob parameter only forks can be configured at any time.

When blobs were first added to the network in the Dencun upgrade, the target was 3 - that was increased to 6 in Pectra and, after Fusaka, that can now be increased at a sustainable rate independently of these major network upgrades.

> [ IMAGE HERE, ETH.ORG TEAM PLEASE HELP ATTACH: https://drive.google.com/file/d/1U24kK-XbZ_l6UWwDO_2-0Mxn4kSkEwVe/view?usp=sharing ]

Reviewer: Will upload the image to this directory... We need to add alt text for this though, so if you could update the following with a description:

```suggestion
![todo: replace this with description of diagram for screen readers](./average-blob-count-per-block.webp)

```

Blob parameter only forks can be set by clients, similarly to other configuration like gas limit. Between major Ethereum upgrades, clients can agree to increase the `target` and `max` blobs to e.g. 9 and 12 and then node operators will update to take part in that tiny fork. These blob parameter only forks can be configured at any time.

When blobs were first added to the network in the Dencun upgrade, the target was 3 - that was increased to 6 in Pectra and, after Fusaka, that can now be increased at a sustainable rate independently of these major network upgrades.

> ![todo: replace this with description of diagram for screen readers](./average-blob-count-per-block.webp)

Reviewer: ```suggestion
![Chart showing average blob count per block and increasing targets with upgrades](./average-blob-count-per-block.webp)
```

- layer 2s pay at least a meaningful slice of the compute they force on nodes
- base-fee spikes on the EL can no longer strand the blob fee at 1 wei

**Resources**:
- [EIP-7918 technical specification](https://eips.ethereum.org/EIPS/eip-7918)
- [Storybook explainer](https://notes.ethereum.org/@anderselowsson/AIG)

### Scale L1 {#scale-l1}

> #### eth/69 - history expiry and simpler receipts {eth-69}

Reviewer: "eth/69"?

If this is a term for something, we should explain it, and unless a specific reason would "Sentence case" this by capitalizing starting letter

Otherwise:

```suggestion
#### History expiry and simpler receipts {#history-expiry}
```

**Resources**: [EIP-7823 technical specification](https://eips.ethereum.org/EIPS/eip-7823)

#### Transaction Gas Limit Cap {#transaction-gas-limit-cap}

EIP-[7825](https://eips.ethereum.org/EIPS/eip-7825) adds a cap of 16,777,216 (2^24) gas per transaction. It’s proactive DoS hardening by bounding the worst-case cost of any single transaction as we raise the block gas limit. It makes validation and propagation easier to model to allow us to tackle scaling via raising the gas limit.

Why exactly 2^24 gas? It’s comfortably smaller than today’s gas limit, is large enough for real contract deployments & heavy precompiles, and a power of 2 makes it easy to implement across clients. This new maximum transaction size is a similar to pre-Pectra average block size, making it a reasonable limit for any operation on Ethereum.

**Resources**: [EIP-7825 technical specification](https://eips.ethereum.org/EIPS/eip-7825)

> #### MODEXP Gas Cost Increase {#modexp-gas-cost-increase}

Reviewer: ```suggestion
#### `MODEXP` gas cost increase {#modexp-gas-cost-increase}
```

By better matching costs to actual processing time, MODEXP can no longer cause a block to take too long to validate. This change is one of several aimed at making it safe to increase Ethereum’s block gas limit in the future.

**Resources**: [EIP-7883 technical specification](https://eips.ethereum.org/EIPS/eip-7883)

#### RLP Execution Block Size Limit {#rlp-execution-block-size-limit}

This creates a ceiling on how big a block is allowed to be - this is a limit on what's *sent* over the network and is separate from the gas limit, which limits the *work* inside a block. The block size cap is 10 MiB, with a small allowance (2 MiB) reserved for consensus data so that everything fits and propagates cleanly. If a block shows up bigger than that, the clients reject it.
This is needed because very large blocks take longer to spread and verify across the network and can create consensus issues or be abused as a DoS vector. Also, the consensus layer’s gossip already won’t forward blocks over ~10 MiB, so aligning the execution layer to that limit avoids weird “seen by some, dropped by others” situations.

> The nitty-gritty: this is a cap on the [RLP](/developers/docs/data-structures-and-encoding/rlp/)-encoded execution block size. 10 MiB total, with a 2 MiB safety margin reserved for beacon-block framing. Practically, clients define

Reviewer: (ref: https://www.grammarbook.com/numbers/numbers.asp, Rule 1)

```suggestion
The nitty-gritty: this is a cap on the [RLP](/developers/docs/data-structures-and-encoding/rlp/)-encoded execution block size. Ten (10) MiB total, with a two (2) MiB safety margin reserved for beacon-block framing. Practically, clients define
```

...or could compromise and slip in another word to start the sentence

`MAX_BLOCK_SIZE = 10,485,760` bytes and 

`SAFETY_MARGIN = 2,097,152` bytes, 

and reject any execution block whose RLP payload exceeds 

`MAX_RLP_BLOCK_SIZE = MAX_BLOCK_SIZE − SAFETY_MARGIN`

> The goal is to bound worst-case propagation/validation time and align with CL gossip behavior, reducing reorg/DoS risk without changing gas accounting.

Reviewer: ```suggestion
The goal is to bound worst-case propagation/validation time and align with consensus layer gossip behavior, reducing reorg/DoS risk without changing gas accounting.
```

**Resources**: [EIP-7935 technical specification](https://eips.ethereum.org/EIPS/eip-7935)

### Improve UX {#improve-ux}

#### Deterministic proposer lookahead {#deterministic-proposer-lookahead}

With EIP-7917, Beacon Chain will become aware of upcoming block proposers for the next epoch. Having a deterministic view on which validators will be proposing future blocks can enable [preconfirmations](https://ethresear.ch/t/based-preconfirmations/17353) - a commitment with the upcoming proposer that guarantees the user transaction will be included in their block without waiting for the actual block.

This feature benefits client implementations and security of the network as it prevents edge cases where validators could manipulate the proposer schedule. The lookahead also allows for less complexity of the implementation.

**Resources**: [EIP-7917 technical specification](https://eips.ethereum.org/EIPS/eip-7917)

#### Count leading zeros (CLZ) opcode {#count-leading-zeros-opcode}

> This feature adds a small EVM instruction, CLZ (“count leading zeros”). Most everything in the EVM is represented as a 256-bit value - this new opcode returns how many zero bits are at the front. This is a common feature in many instruction set architectures as it enables more efficient arithmetic operations. In practice this collapses today’s hand-rolled bit scans into one step, so finding the first set bit, scanning bytes, or parsing bitfields becomes simpler and cheaper. The opcode is low, fixed-cost and has been benchmarked to be on par with a basic add, which trims bytecode and saves gas for the same work.

Reviewer: (ref: https://ethereum.org/contributing/style-guide/#acronyms)

```suggestion
This feature adds a small EVM instruction, **count leading zeroes (CLZ)**. Most everything in the EVM is represented as a 256-bit value—this new opcode returns how many zero bits are at the front. This is a common feature in many instruction set architectures as it enables more efficient arithmetic operations. In practice this collapses today’s hand-rolled bit scans into one step, so finding the first set bit, scanning bytes, or parsing bitfields becomes simpler and cheaper. The opcode is low, fixed-cost and has been benchmarked to be on par with a basic add, which trims bytecode and saves gas for the same work.
```

Introduces a built-in, passkey-style secp256r1 (P-256) signature checker at the fixed address `0x100` using the same call format already adopted by many L2s and fixing edge cases, so contracts written for those environments work on L1 without changes.

UX upgrade! For users, this unlocks device-native signing and passkeys. Wallets can tap into Apple Secure Enclave, Android Keystore, hardware security modules (HSMs), and FIDO2/WebAuthn directly - no seed phrase, smoother onboarding, and multi-factor flows that feel like modern apps. This results in better UX, easier recovery, and account abstraction patterns that match what billions of devices do already.

For developers, it takes a 160-byte input and returns a 32-byte output, making it easy to port existing libraries and L2 contracts. Under the hood, it includes point-at-infinity and modular-comparison checks to eliminate tricky edge cases without breaking valid callers.

**Resources**: 
- [EIP-7951 technical specification](https://eips.ethereum.org/EIPS/eip-7951)
- [More about RIP-7212](https://www.alchemy.com/blog/what-is-rip-7212) _(Note that EIP-7951 superseded RIP-7212)_

### Meta {#meta}

> #### `eth_config` JSON-RPC Method {#eth-config}

Reviewer: (ref: https://ethereum.org/contributing/style-guide/content-standardization/#header-casing)

```suggestion
#### `eth_config` JSON-RPC method {#eth-config}
```

This is a JSON-RPC call that allows you to ask your node what fork settings you're running. It returns three snapshots: `current`, `next`, & `last` so that validators and monitoring tools can verify that clients are lined up for an upcoming fork.

> Practically speaking, this is to address a shortcoming discovered when the Pectra fork went live on the Holešky testnet in early 2025 with minor misconfigurations which resulted in a non-finalizing state. This helps testing teams and developers ensure that major forks will behave as expected when moving from devnets to testnets and from testnets to mainnet.

Reviewer: (site has standardized with the "Holesky" spelling for English pages, and other ref: https://ethereum.org/contributing/style-guide/content-standardization/#mainnet)

```suggestion
Practically speaking, this is to address a shortcoming discovered when the Pectra fork went live on the Holesky testnet in early 2025 with minor misconfigurations which resulted in a non-finalizing state. This helps testing teams and developers ensure that major forks will behave as expected when moving from devnets to testnets, and from testnets to Mainnet.
```

Snapshots include: `chainId`, `forkId`, planned fork activation time, which precompiles are active, precompile addresses, system contract dependencies, and the fork's blob schedule.

> This EIP is in a section apart from the "Core EIPs" because the fork doesn't actually implement any changes - it's a notice that client teams must implement this JSON-RPC method by the Fusaka upgrade.

Reviewer: ```suggestion
This EIP is in a section apart from the "Core EIPs" because the fork doesn't actually implement any changes—it's a notice that client teams must implement this JSON-RPC method by the Fusaka upgrade.
```
